front50-s3/src/main/java/com/netflix/spinnaker/front50/model/EventingS3ObjectKeyLoader.java

```diff
@@ -0,0 +1,273 @@
+              () -> {
+                log.debug("Refreshing object keys for {} (asynchronous)", objectType);
+                return s3StorageService.listObjectKeys(objectType);
+              }
+            );
+            executor.execute(task);
+            return task;
+          }
+        }
+      );
+
+    this.rootFolder = s3Properties.getRootFolder();
+
+    if (scheduleImmediately) {
+      taskScheduler.schedule(this, new Date());
+    }
+  }
+
+  @Override
+  @PreDestroy
+  public void shutdown() {
+    log.debug("Stopping ...");
+    pollForMessages = false;
+    log.debug("Stopped");
+  }
+
+  @Override
+  public Map<String, Long> listObjectKeys(ObjectType objectType) {
+    try {
+      Map<String, Long> objectKeys = objectKeysByObjectTypeCache.get(objectType);
+      objectKeysByLastModifiedCache.asMap().entrySet()
+        .stream()
+        .filter(e -> e.getKey().objectType == objectType)
+        .forEach(e -> {
+          String key = e.getKey().key;
+          if (objectKeys.containsKey(key)) {
+            Long currentLastModifiedTime = e.getValue();
+            Long previousLastModifiedTime = objectKeys.get(key);
+            if (currentLastModifiedTime > previousLastModifiedTime) {
+              log.info(
+                "Detected Recent Modification (key: {}, previous: {}, current: {})",
+                key,
+                new Date(previousLastModifiedTime),
+                new Date(e.getValue())
+              );
+              objectKeys.put(key, currentLastModifiedTime);
+            }
+          }
+        });
+      return objectKeys;
+    } catch (ExecutionException e) {
+      log.error("Unable to fetch keys from cache", e);
+      return s3StorageService.listObjectKeys(objectType);
+    }
+  }
+
+  @Override
+  public void run() {
+    while (pollForMessages) {
+      List<Message> messages = temporarySQSQueue.fetchMessages();
+
+      if (messages.isEmpty()) {
+        continue;
+      }
+
+      messages.forEach(message -> {
+        S3Event s3Event = unmarshall(objectMapper, message.getBody());
+        if (s3Event != null) {
+          tick(s3Event);
+        }
+        temporarySQSQueue.markMessageAsHandled(message.getReceiptHandle());
+      });
+    }
+  }
+
+  private void tick(S3Event s3Event) {
+    s3Event.records.forEach(record -> {
+      if (record.s3.object.key.endsWith("last-modified.json")) {
+        return;
+      }
+
+      String eventType = record.eventName;
+      KeyWithObjectType keyWithObjectType = buildObjectKey(rootFolder, record.s3.object.key);
+      DateTime eventTime = new DateTime(record.eventTime);
+
+      log.debug(
+        "Received Event (objectType: {}, type: {}, key: {}, delta: {})",
+        keyWithObjectType.objectType,
+        eventType,
+        keyWithObjectType.key,
+        System.currentTimeMillis() - eventTime.getMillis()
+      );
+
+      objectKeysByLastModifiedCache.put(keyWithObjectType, eventTime.getMillis());
```

Comment:
Seemed easier to avoid concurrency issues if we tracked detected modifications and applied when `listObjectKeys()` was called ... vs trying to apply them directly to the cache as changes came in.

From testing I do not believe that the rate of change is going to be overwhelming.